#### Imbalanced dataset

A dataset with an unequal class distribution

In [1]:
#lets import libraries
import pandas as pd
import numpy as np

In [2]:
#loading dataset to pandas dataframe
credit_data = pd.read_csv(r"C:\Users\bhara\Downloads\credit_data.csv")

In [3]:
#printing first 5 rows of the dataset
credit_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
#checking the dataset total rows and columns
credit_data.shape

(284807, 31)

In [5]:
#checking the missing values
credit_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
#determine the distribution of two class
credit_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Here, we can see the 0 --> legit transaction has 284315

and 0 ----> fraud transaction has 492 wherein here we can see the data is imbalanced

### How to handled imbalance dataset

#### Separating the legit and fraudulent transactions

In [8]:
legit = credit_data[credit_data.Class == 0]
fraud = credit_data[credit_data.Class == 1]

In [9]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


### We are using under sampling method 

Building a sample dataset containing similiar distribution of Legit and Fraudulent transaction

Number of Fraudulent transactions  -- > 492

In [10]:
legit_sample = legit.sample(n=492)

In [11]:
print(legit_sample.shape)

(492, 31)


#### Concatenate the two dataframes

In [13]:
new_dataset = pd.concat([legit_sample, fraud], axis = 0)

In [17]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
99380    67092.0  1.340308 -0.689120 -0.316892 -0.986277 -0.370683 -0.165933   
277705  167809.0  2.005758  0.023493 -2.130075  0.075374  0.846818 -0.061499   
195768  131212.0  1.727854 -0.705376 -0.764415  0.614702 -0.633329 -0.618949   
265106  161744.0 -1.579821 -1.171667  0.851443  0.291326  2.444016 -1.454168   
219599  141795.0 -0.981369 -0.675014  1.184065  0.208224 -1.259620  0.711967   

              V7        V8        V9  ...       V21       V22       V23  \
99380  -0.405386 -0.086345 -1.086664  ...  0.226674  0.416047 -0.328144   
277705  0.055240 -0.056239  0.651399  ... -0.240509 -0.496529  0.067069   
195768 -0.179811 -0.055690  0.672166  ...  0.132703  0.196791  0.103046   
265106 -0.518872  0.164331 -0.567994  ... -0.037176 -0.650636  0.208072   
219599  0.063534  0.274459 -0.352875  ... -0.445827 -0.538165  0.120446   

             V24       V25       V26       V27       V28  Amount  Class  
99380  -0.799051  0.733539 -0.068364 -0.023345 -0.000219   75.95      0  
277705 -0.248782  0.067836 -0.093757 -0.012418 -0.023887   32.49      0  
195768  0.089885 -0.313328  0.225749 -0.066654 -0.040968  132.00      0  
265106 -0.469639 -0.207287  0.400128 -0.001422  0.148165    1.79      0  
219599  1.057497 -0.361940 -0.914954 -0.118400 -0.452530  222.39      0  

[5 rows x 31 columns]

In [18]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [19]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [21]:
new_dataset.shape

(984, 31)